plugins/catalog-backend-module-gitlab/src/lib/client.ts

```diff
@@ -428,20 +413,39 @@ export class GitLabClient {
       }
     }
 
-    const response = await fetch(
-      request.toString(),
-      getGitLabRequestOptions(this.config),
-    );
-
-    if (!response.ok) {
-      throw new Error(
-        `Unexpected response when fetching ${request.toString()}. Expected 200 but got ${
-          response.status
-        } - ${response.statusText}`,
+    for (let i = 0; ; i++) {
+      this.logger.debug(`Fetching: ${request.toString()}`);
+      const response = await fetch(
+        request.toString(),
+        getGitLabRequestOptions(this.config),
       );
-    }
 
-    return response.json();
+      if (response.status === tooManyRequests && i < maxRetry) {
+        const retryAfter = response.headers.get('Retry-After');
+
+        if (retryAfter) {
+          const delay = isNaN(Number(retryAfter))
+            ? new Date(retryAfter).getTime() - Date.now() // Retry-After is a date
```

Comment:
Should we be having a `maxDelay` or something here, it feels a little unfortunate for consumers of this API to not fail straight away and potentially block anything more than a few seconds to be honest. 

Take the catalog for instance, if you have 1000 components and all those reads are delayed 15minutes, you're going to lock up the catalog processing waiting for these reads to resolve, rather than failing straight away and it being re-processed at a later time.

Reply:
I'll update to satisfy the bot and maybe we can discuss at an upcoming SIG.